### ISSUE TITLE
Installer doesn't work behind proxy

### ISSUE THREAD

**Issue by kappa7194** [git-for-windows/git#46]
Ported from https://github.com/msysgit/git/issues/336:

I downloaded the `git-sdk-64-installer-preview.7z.exe` installer from https://github.com/git-for-windows/build-extra/releases/tag/net-installer-preview-1, decompressed its content to a directory, and then ran the `setup-git-sdk.bat` script.

The script fails immediately with the following output:

```
:: Synchronizing package databases...
 git-for-windows                                                                                  0.0   B  0.00B/s 00:00 [------------------------------------------------------------------------]   0%
 git-for-windows                                                                               1392.0   B  1359K/s 00:00 [########################################################################] 100%

error: failed retrieving file 'mingw32.db' from downloads.sourceforge.net : Connection timed out after 10000 milliseconds
error: failed to update mingw32 (download library error)
error: failed retrieving file 'mingw64.db' from downloads.sourceforge.net : Connection timed out after 10000 milliseconds
error: failed to update mingw64 (download library error)
error: failed retrieving file 'msys.db' from downloads.sourceforge.net : Connection timed out after 10000 milliseconds
error: failed to update msys (download library error)
error: failed to prepare transaction (could not find database)
warning: database file for 'mingw32' does not exist
warning: database file for 'mingw64' does not exist
warning: database file for 'msys' does not exist
error: target not found: pacman
warning: database file for 'mingw32' does not exist
warning: database file for 'mingw64' does not exist
warning: database file for 'msys' does not exist
:: There is 1 member in group base:
:: Repository git-for-windows
   1) msys2-runtime

Enter a selection (default=all):
error: target not found: python
error: target not found: less
error: target not found: openssh
error: target not found: patch
error: target not found: make
error: target not found: tar
error: target not found: diffutils
error: target not found: ca-certificates
error: target not found: perl-Error
error: target not found: perl
error: target not found: perl-Authen-SASL
error: target not found: perl-libwww
error: target not found: perl-MIME-tools
error: target not found: perl-Net-SMTP-SSL
error: target not found: perl-TermReadKey
error: target not found: mintty
error: target not found: mingw-w64-x86_64-toolchain
error: target not found: mingw-w64-x86_64-curl
error: target not found: mingw-w64-x86_64-expat
error: target not found: mingw-w64-x86_64-openssl
error: target not found: mingw-w64-x86_64-tcl
error: target not found: mingw-w64-x86_64-pcre
'mintty' is not recognized as an internal or external command,
operable program or batch file.
```

After retrying about a dozen times to rule out problems with SourceForge (as [suggested by dscho]) I switched the computer from the corporate network to the hotspot created by my mobile phone the script succeeded on the first try.

It seems the `setup-git-sdk.bat` script doesn't support proxies, as our corporate network doesn't have direct internet access — everything must pass through an HTTP proxy — while the mobile phone hotspot has.

  [suggested by dscho]: https://github.com/msysgit/git/issues/336#issuecomment-86478265

**Comment by kappa7194** [comment]
Interestingly, when the installer [clones Git's repository] it uses the proxy: when using the mobile phone hotspot I forgot to remove the proxy configuration from Windows, and the clone failed because it couldn't contact the remote, after removing the proxy configuration from Windows the clone succeeded.

  [clones Git's repository]: https://cloud.githubusercontent.com/assets/1249988/6851098/42c5de4a-d3dd-11e4-972f-6fcb2f17937c.png

**Comment by dscho** [comment]
Okay, I have no way to reproduce, so I rely on you to figure out the solution (after all the net installer is a *developers*' preview).

Could you please keep the window open after `pacman` fails? You will find a file called `setup-git-sdk.bat` in the top-level Git SDK directory. This is the batch file that is executed to initialize the Git SDK.

Now, what I need to know is: what is the `%ERRORLEVEL%` if you change working directory in a new `cmd` window to the top-level Git SDK directory and then run `usr\bin\pacman -Sy`?

This is the first thing we need to find out, i.e. a method to figure out *reliably* that a proxy is required.

After that, we need to ask the user for a proxy, then set the environment variable `http_proxy` (make sure to use all lower-case). This assumes, of course, that `pacman` respects that variable (could you please find that out, too? It's just calling `set http_proxy=...` and then `usr\bin\pacman -Sy`).

Once that is all said and done, we should add an `@IF NOT "%http_proxy%" == "" @ECHO "export http_proxy=%http_proxy%" >etc\profile.d\http-proxy.sh` line near the end of the setup script.

**Comment by dscho** [comment]
> After that, we need to ask the user for a proxy, then set the environment variable http_proxy (make sure to use all lower-case). This assumes, of course, that pacman respects that variable (could you please find that out, too? It's just calling set http_proxy=... and then usr\bin\pacman -Sy).

According to Pacman's documentation, this method should work: http://www.garron.me/en/go2linux/arch-linux-pacman-proxy.html

**Comment by nalla** [comment]
You have to set `http_proxy` and `https_proxy` environment variables. This is case sensetive.

**Comment by nalla** [comment]
You have to set *both* proxy variables. `http_proxy` and `https_proxy`.

**Comment by nalla** [comment]
@dscho you do not need to `export` the `%http[s]_proxy` variables. They *should* be catched up from the windows environment seamlessly from `pacman`. At least that was the case for me. We may have to provide some install instructions for the `net-installer`?

BTW: I struggled with this as well and forgot to put in on the wiki. Sorry for that!

**Comment by dscho** [comment]
@nalla I think we should export those variables from a `profile.d/` script. That way, the user has to type in the HTTP proxy only once, and subsequent Git SDK sessions will work, too.

I have to admit, though, that I would have expected all my hard work to be honored by more action on the bug reporter's part than with silence.

**Comment by dscho** [comment]
@kappa7194 hmm. I would have expected somewhat better communication from you. Hopefully I can at least interest you in running my *attempt* at fixing your problem: https://github.com/git-for-windows/build-extra/releases/download/issue-46/git-sdk-64-installer-issue-46.7z.exe

**Comment by nalla** [comment]
A `profile.d` script would indeed help future sessions. If I remember correctly I put these variables manually into the environment via control panel. That was the reason I needed no further script. Lets just ask the user then if he has proxy settings when running the setup script. Maybe print out the location of the created script too if he had a typo?

**Comment by dscho** [comment]
@nalla precisely: https://github.com/git-for-windows/build-extra/commit/net-installer-proxy

**Comment by kappa7194** [comment]
>You have to set both proxy variables. http_proxy and https_proxy.

Manually setting `http_proxy` in a console before invoking `setup-git-sdk.bat` worked. I had to set the variable with `set` and not `setx` (see http://superuser.com/questions/79612/setting-and-getting-windows-environment-variables-from-the-command-prompt) as the latter didn't work.

>I have to admit, though, that I would have expected all my hard work to be honored by more action on the bug reporter's part than with silence.

>@kappa7194 hmm. I would have expected somewhat better communication from you. Hopefully I can at least interest you in running my attempt at fixing your problem

Are you aware that there are other timezones than your own can have lives besides the internet? Moreover I can reproduce the issue only on the corporate network since I don't have a proxy server at home, and that means an eight hours a day window.

**Comment by kappa7194** [comment]
The setup still fails.

When the second console window opens, the one which clones and builds Git, at the beginning I get a couple of errors:

    bash: /mingw64/bin/git: Bad address
    bash: /mingw64/bin/git: Bad address
    bash: /mingw64/bin/git: Bad address

Git's compilation however seems to complete successfully, as that console at the end returns to the prompt.

Then the setup gets stuck: the console where Git has been built return to the prompt and stays there waiting for input, and the original console where `setup-git-sdk.bat` has been launched seems to be waiting for the other one to complete.

If I close the second console (<kbd>CTRL-D</kbd>, logout, etc) the first one starts throwing tons of errors, I haven't been able to capture them all as they quickly filled the console buffer, they were mainly "command not found" errors.

What I've been able to intercept is:

    mkdir: cannot change permissions of '/dev/shm': Permission denied
    mkdir: cannot change permissions of '/dev/mqueue': Permission denied
    'C:\\Windows\\System32\\drivers\\etc\\hosts' -> '/etc/hosts'
    'C:\\Windows\\System32\\drivers\\etc\\protocol' -> '/etc/protocols'
    'C:\\Windows\\System32\\drivers\\etc\\services' -> '/etc/services'
    'C:\\Windows\\System32\\drivers\\etc\\networks' -> '/etc/networks'
    bash: /mingw64/bin/create-shortcut.exe: Bad address

Trying to rerun the setup yielded to something worse:

    bash: line 1:  5768 Aborted                 (core dumped) git clone -b master -c core.autocrlf=false https://github.com/git-for-windows/git

Is there a way to redirect the whole output of _both_ console windows to a file while still showing it in the console? I've found ` cmd > file.txt 2 > &1` but this hides the output from the console leaving you unable to understan what's going on.

At the moment I'm deleting the directory where the setup lives to start with a clean copy, is there a way to avoid re-downloading the libraries every time while still trying in a pristine environment? Our sysadmins will start to get angry if I try too much times as we share the network connection with the production servers…
